### PR TITLE
Refactor MatrixRead and MatrixReader to be cleaner

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: src/main/scala/is/hail/expr/Parser.scala [hunk 1]
```diff
@@ -819,11 +819,11 @@ object Parser extends JavaTokenParsers {
...
-      "MatrixRange" ~> int32_literal ~ int32_literal ~ int32_literal_opt ~ boolean_literal ~ boolean_literal ^^ { case nRows ~ nCols ~ nPartitions ~ dropCols ~ dropRows =>
-        MatrixIR.range(HailContext.get, nRows, nCols, nPartitions, dropCols, dropRows)
-      } |
-      "MatrixRead" ~> string_literal ~ boolean_literal ~ boolean_literal ~ matrix_type_expr_opt ^^ { case path ~ dropCols ~ dropRows ~ requestedType =>
-        MatrixIR.read(HailContext.get, path, dropCols, dropRows, requestedType)
+      "MatrixRead" ~> matrix_type_expr_opt ~ boolean_literal ~ boolean_literal ~ string_literal ^^ {
+        case typ ~ dropCols ~ dropRows ~ readerStr =>
+          implicit val formats = ir.MatrixReader.formats
+          val reader = Serialization.read[ir.MatrixReader](readerStr)
+          ir.MatrixRead(typ, dropCols, dropRows, reader)
```
Comment (cseed) [inline src/main/scala/is/hail/expr/Parser.scala]: I personally would make MatrixRead take a non-optional type and pass in `typ.getOrElse(reader.typ)` here, I think that's a tad simpler.  But this is also fine.

File: src/main/scala/is/hail/expr/ir/Pretty.scala [hunk 2]
```diff
@@ -190,19 +182,11 @@ object Pretty {
...
-                  "None"
-                else
-                  typ.parsableString())
-            case MatrixRead(typ, _, _, dropCols, dropRows, MatrixRangeReader(nRows, nCols, nPartitions)) =>
-              s"$nRows $nCols" + " " +
-                nPartitions.map(_.toString).getOrElse("None") + " " +
-                prettyBooleanLiteral(dropCols) + " " +
-                prettyBooleanLiteral(dropRows)
+            case MatrixRead(typ, dropCols, dropRows, reader) =>
+              typ.getOrElse(reader.fullType).parsableString() + " " +
```
Comment (cseed) [inline src/main/scala/is/hail/expr/ir/Pretty.scala]: Actually, shouldn't this output None for the type if it is None?

Comment (tpoterba) [reply src/main/scala/is/hail/expr/ir/Pretty.scala]: ahh yes

Comment (tpoterba) [reply src/main/scala/is/hail/expr/ir/Pretty.scala]: will fix

File: python/hail/ir/matrix_ir.py [hunk 3]
```diff
@@ -18,8 +20,12 @@ def __init__(self, path, drop_cols, drop_rows):
         self.drop_rows = drop_rows
 
     def __str__(self):
-        return '(MatrixRead "{}" {} {} None)'.format(
-            self.path, self.drop_cols, self.drop_rows)
+        config = dict(
+            name='MatrixNativeReader',
+            path=self.path
+        )
+        return f'(MatrixRead None {self.drop_cols} {self.drop_rows} "{escape_str(json.dumps(config))})"'
```
Comment (cseed) [inline python/hail/ir/matrix_ir.py]: Paren on the wrong side of the " here.

**Pull request merged**